### ISSUE TITLE
Cosine_annealing does not work

### ISSUE THREAD

**Issue by BigFishMaster** [OpenNMT/OpenNMT-tf#378]
I train gpt2 use the following parameters, but cosine_annealing does not work after the warmup step.
```python
params:
    average_loss_in_time: true
    label_smoothing: 0.1
    optimizer: AdamOptimizer
    optimizer_params:
        beta1: 0.9
        beta2: 0.998
    weight_decay: 0.01
    learning_rate: 2.5e-4
    decay_type: cosine_annealing
    decay_params:
        max_step: 10000000
        warmup_steps: 8000
    gradients_accum: 1
    beam_width: 4
    length_penalty: 0.6
    minimum_decoding_length: 0
    maximum_iterations: 256
```

![image](https://user-images.githubusercontent.com/4915869/54086189-2eb60100-4382-11e9-8e71-e2b4e7eb3f00.png)

**Comment by BigFishMaster** [comment]
Will it be caused by the max_step=10000000?

**Comment by guillaumekln** [comment]
The learning rate will reach 0 when the step is `max_step`. So if this value is very large, it is a very slow decay.

**Comment by BigFishMaster** [comment]
@guillaumekln But I didn't see any decay of the learning rate.
![image](https://user-images.githubusercontent.com/4915869/54122312-a5f99c80-4437-11e9-96a5-5fdf5319faf3.png)

**Comment by guillaumekln** [comment]
You should probably reduce `max_step`. With your configuration, the learning rate is stiil at `0.00024999448` after 30k steps and TensorBoard can not render the difference.

See the decay formula here:

https://github.com/OpenNMT/OpenNMT-tf/blob/v1.21.5/opennmt/utils/decay.py#L96-L117

**Comment by guillaumekln** [comment]
Closing this one. The decay should work as expected.

**Comment by BigFishMaster** [comment]
Yes the decay worked.
